File: class3/exercises/fourSum.py
class Solution:
    """
    @param numbers: Give an array
    @param target: An integer
    @return: Find all unique quadruplets in the array which gives the sum of zero
    """
    def fourSum(self, numbers, target):
        # write your code here
        
        nums = numbers
        
        results = [] 
        
        if nums == None or len(nums) < 4:
            return results 
            
        nums.sort()
            
        for i in range(0, len(nums) - 3):
            
            if i and nums[i] == nums[i - 1]:
                continue 
            
            for j in range(i + 1, len(nums) - 2):
                
                if (j - 1) > i and nums[j] == nums[j - 1]:
                    continue 
                
                left, right = j + 1, len(nums) - 1 
                
                half_target = target - nums[i] - nums[j]
                
                while left < right:
                    
                    if nums[left] + nums[right] < half_target:
                        left += 1 
                        
                    elif nums[left] + nums[right] == half_target:
                        results.append( [ nums[i], nums[j], nums[left], nums[right] ] )
                        
                        left += 1 
                        right -= 1 
                        
                        while left < right and nums[left] == nums[left - 1]:
                            left += 1 
                        
                        while left < right and nums[right] == nums[right + 1]:
                            right -= 1 
                        
                    elif nums[left] + nums[right] > half_target:
                        right -= 1 
                        
        return results

File: interviews/Twitch/JavaTwitch/FourSum.java
import java.util.*;
public class FourSum {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        
        List<List<Integer>> results = new ArrayList<List<Integer>>();
        
        if (nums == null || nums.length < 4) return results;
        
        Arrays.sort(nums);
        
        for (int i = 0; i < nums.length - 3; i++) {
            
            if (i > 0 && nums[i] == nums[i - 1]) continue; 
            
            for (int j = i + 1; j < nums.length - 2; j++) {
                
                if (j > i + 1 && nums[j] == nums[j - 1]) continue; 
                
                int start = j + 1, end = nums.length - 1; 
                int targetSum = target - nums[i] - nums[j]; 
                
                while (start < end) {
                    
                    if (nums[start] + nums[end] < targetSum) {
                        start += 1;
                    }
                    
                    else if (nums[start] + nums[end] > targetSum) {
                        end -= 1;
                    }
                    
                    else {
                        
                        List<Integer> result = new ArrayList<Integer>(); 
                        result.add(nums[i]);
                        result.add(nums[j]);
                        result.add(nums[start]);
                        result.add(nums[end]);
                        
                        results.add(result);
                        
                        start += 1; 
                        while (start < end && nums[start] == nums[start - 1]) start += 1;
                        
                        end -= 1;
                        while (start < end && nums[end] == nums[end + 1]) end -= 1;
                    }
                }
            }
            
            
        }
        
        return results;
        
    }
}